llvm/projects/evm-sdk/evm-stdlib/include/storage.h

```diff
@@ -0,0 +1,256 @@
+#include "evm_sdk.h"
+#include <utility>
+
+namespace evm::stor {
+
+namespace details {
+
+template <unsigned... Slots>
+void remove_sequence(uint256_t slot,
+                     std::integer_sequence<unsigned, Slots...> seq) {
+  (__builtin_evm_sstore(slot + Slots, 0), ...);
+}
+
+template <typename T> void clear(uint256_t slot) {
+  if constexpr (std::is_integral_v<T>) {
+    __builtin_evm_sstore(slot, 0);
+  } else {
+    remove_sequence(slot,
+                    std::make_integer_sequence<unsigned, T::FIELDS_NUM>{});
+  }
+}
+}  // namespace details
+
+template <class T, unsigned slot_>
+struct SlotStaticImpl {
+  using ValueType = T;
+
+  static uint256_t slot() {
+    return slot_;
+  }
+};
+
+template <class T>
+struct SlotImpl {
+  using ValueType = T;
+
+  explicit SlotImpl(uint256_t slot) : slot_(slot) {}
+
+  uint256_t slot() const {
+    return slot_;
+  }
+
+  uint256_t slot_;
+};
+
+template <class Impl>
+struct SlotBase: private Impl {
+  using ValueType = typename Impl::ValueType;
+  using Impl::Impl;
+
+  void set(const ValueType &v) {
+    __builtin_evm_sstore(Impl::slot(), v);
+  }
+
+  ValueType get() const {
+    return __builtin_evm_sload(Impl::slot());
+  }
+
+  SlotBase& operator=(const ValueType &v) {
+    set(v);
+    return *this;
+  }
+
+  operator ValueType() const {
+    return get();
+  }
+};
+
+template<typename T>
+using Slot = SlotBase<SlotImpl<T>>;
+
+template<typename T, uint256_t Key>
+using SlotStatic = SlotBase<SlotStaticImpl<T, Key>>;
+
+/*******************************************************************************
+ * Dynamic array.
+ */
+template<typename Impl>
+class VectorBase: private Impl {
+public:
+  using ValueType = typename Impl::ValueType;
+  using Impl::Impl;
+
+  auto operator[](uint256_t k) {
+    if constexpr (std::is_integral_v<ValueType>) {
+      return Slot<ValueType>(get_storage_key() + k);
+    } else {
+      return ValueType(get_storage_key() + k * ValueType::FIELDS_NUM);
+    }
+  }
+
+  auto operator[](uint256_t k) const {
+    return const_cast<VectorBase*>(this)->operator[](k);
+  }
+
+  unsigned size() const {
+    return __builtin_evm_sload(Impl::slot());
+  }
+
+  bool empty() const {
+    return size() == 0;
+  }
+
+  auto push() {
+    auto sz = size();
+    uint256_t new_size = sz + 1;
+    __builtin_evm_sstore(Impl::slot(), new_size);
+    return operator[](sz);
+  }
+
+  void pop() {
+    auto sz = size();
+    require(sz != 0);
+    details::clear<ValueType>(get_storage_key() + sz - 1);
+    __builtin_evm_sstore(Impl::slot(), sz - 1);
+  }
+
+private:
+  uint256_t get_storage_key() const {
+    return __builtin_evm_sha3_vargs(Impl::slot());
+  }
+};
+
+template<typename T>
+using Vector = VectorBase<SlotImpl<T>>;
+
+template<typename T, uint256_t Slot>
+using VectorStatic = VectorBase<SlotStaticImpl<T, Slot>>;
```

Comment:
Doesn't this expose `push()` and `pop()` for a fixed-size array?

I also wonder in this and other places where you use private inheritance from a template parameter. Does this have a specific reason? When reading the code it wasn't very obvious to me at the first glance.

Reply:
> Doesn't this expose `push()` and `pop()` for a fixed-size array?

Static doesn't mean that Vector is a fixed size. It means that the slot(or storage key) is known at compile time, and it is never changed. Non-static version, in turn, doesn't know where it will be placed in the storage, thus it has runtime variable for this.
For example:
```c++
struct Foo {
   Vector<int> vec2;
}
Map<Foo> map;
Vector<Foo> vec1;
```
Here, `vec1` is static, and its size is stored in slot 1 (it is sequential number of vec1) and data is stored in slot started from `sha3(1)`. `vec2` is non-static, and its size is stored in slot `sha3(1, key)` and data is stored in `sha3(sha3(1, key))`.

> I also wonder in this and other places where you use private inheritance from a template parameter. Does this have a specific reason? When reading the code it wasn't very obvious to me at the first glance.

Private inheritance is used just for hiding the interface of Impl classes.

Reply:
Oh I see. This answers my question then. I just didn't understand what a static vs dynamic is.